### PR TITLE
feat(console): User can add a home page for a v4 API: Choose Existing…

### PR DESCRIPTION
**Pull request by illichso**
… Page for Homepage

## Issue

https://gravitee.atlassian.net/browse/APIM-5923

## Description

As an API publisher

I want to choose a home page from existing pages for my v4 API

so that users in the portal can see the home page under the Details section when they find my API in the catalog.

## Additional context

https://github.com/user-attachments/assets/c4cb9474-2f5c-48da-8717-8396b5aa7f9a

Screenshot 2024-08-29 at 18 04 18

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-usbjiekggg.chromatic.com)

### CODE REVIEW

File: gravitee-apim-console-webui/src/management/api/documentation-v4/api-documentation-v4.module.ts [hunk 1]
```diff
@@ -121,6 +123,8 @@ import { GioApiMetadataListModule } from '../component/gio-api-metadata-list/gio
     ApiDocumentationV4BreadcrumbComponent,
     ApiDocumentationV4FileUploadComponent,
     ApiDocumentationV4PageTitleComponent,
+    ApiDocumentationChoosePageListComponent,
+    ApiDocumentationChooseExistingPageComponent,
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/api-documentation-v4.module.ts]: Why do you need to declare them in the module? If they are standalone, I believe that they don't need to be included in any modules.

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    <mat-dialog-actions class="exit-without-save" align="end">
+      <button mat-raised-button [disabled]="isReadOnly" (click)="goBackToDefaultPage()">Exit without saving</button>
+    </mat-dialog-actions>
+  </div>
+
+  <mat-dialog-content class="custom-dialog-content">
+    <api-documentation-choose-page-list [pages]="pages" (selectPage)="selectPage($event)"> </api-documentation-choose-page-list>
+  </mat-dialog-content>
+  <gio-banner-warning class="gio-banner-warning-content">
+    This choice is final, and once a page is set as the homepage, it will no longer be listed under Documentation Pages.
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.html]:
```suggestion
    This choice is final. Once a page is set as the homepage, it will no longer be listed under Documentation Pages.
```

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.scss [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+:host {
+  @include gio-layout.gio-responsive-content-container;
+}
+
+.header {
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  padding: 20px 40px;
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.scss]: Instead of giving every item padding 40x, try using flex + gap. This is best practice because if you ever need to add a new item to the component, you will automatically have the spacing that you need.

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  const init = async (pages: Page[], breadcrumb: Breadcrumb[]) => {
+    await TestBed.configureTestingModule({
+      imports: [NoopAnimationsModule, MatIconTestingModule, GioTestingModule, ApiDocumentationChooseExistingPageComponent],
+      providers: [
+        {
+          provide: ApiDocumentationV2Service,
+          useValue: {
+            getApiPages: jest.fn(),
+          } as unknown as jest.Mocked<ApiDocumentationV2Service>,
+        },
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.spec.ts]: Why not just use the expect http calls?

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+          title: 'Confirm Homepage Selection',
+          content: 'This choice is final. Once a page becomes the homepage, it will no longer be accessible in your Documentation Pages.',
+          confirmButton: 'Confirm',
+        },
+      })
+      .afterClosed()
+      .pipe(filter((confirmed) => !!confirmed))
+      .subscribe({
+        next: (_) => {
+          this.save();
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.ts]: The calls within `this.save()` should be in the main thread after `filter((confirmed) => !!confirmed)` instead of in a separate observable thread.

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+          this.apiDocumentationV2Service.updateDocumentationPage(this.api.id, chosenPage.id, { ...chosenPage, homepage: true }),
+        ),
+      )
+      .subscribe({
+        next: () => {
+          this.snackBarService.success('Homepage chosen successfully');
+          this.goBackToDefaultPage();
+        },
+        error: (error) => {
+          this.snackBarService.error(error?.error?.message ?? 'Error while choosing a homepage');
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.ts]:
```suggestion
          this.snackBarService.error(error?.error?.message ?? 'An error occurred when choosing a homepage');
```

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.scss [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+.documentation-pages-list {
+  &__tree {
+    &__node {
+      display: flex;
+      width: 100%;
+      cursor: pointer;
+      gap: 8px;
+      align-items: center;
+      border-bottom: 1px solid #d3d5dc;
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.scss]: Replace hexcode with theme color

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('should render header with text "Name"', () => {
+    const headerElement = fixture.debugElement.query(By.css('.header'));
+    expect(headerElement).toBeTruthy();
+    expect(headerElement.nativeElement.textContent).toContain('Name');
+  });
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.spec.ts]: Not the most interesting test 😅 You can focus on testing the Acceptance Criteria or edge cases.

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+  it('should render mat-tree', () => {
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.spec.ts]: Instead of using technical language (`mat-tree`), it should speak to the business rules of the component. The same goes for the tests below.

```suggestion
  it('should render the page list', () => {
```

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.ts
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.ts:1]: A user should not be able to choose a page that is used as Terms and Conditions.

Either the page should be not shown, or it should be disabled.

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/documentation-home-page-header/api-documentation-v4-home-page-header.component.html [hunk 10]
```diff
@@ -17,9 +17,17 @@
...
+    *gioPermission="{ anyOf: ['api-documentation-u'] }"
+  >
+    Choose Existing Page
+  </button>
   <api-documentation-v4-add-page-button
-    [disabled]="hasPages"
+    [disabled]="hasHomepage || isReadOnly"
     (addPage)="addPage.emit($event)"
     text="Create New Page"
   ></api-documentation-v4-add-page-button>
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/documentation-home-page-header/api-documentation-v4-home-page-header.component.html]: If a homepage is set, the buttons are no longer shown.

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.html [hunk 11]
```diff
@@ -18,14 +18,16 @@
 <mat-card>
   <api-documentation-home-page-header
     [isReadOnly]="isReadOnly"
-    [hasPages]="page != null"
+    [hasHomepage]="homepage !== undefined"
+    [noCustomPages]="pages === undefined || pages.length === 0"
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.html]:
```suggestion
    [hasHomepage]="!!homepage"
    [noCustomPages]="!pages?.length"
```

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.html [hunk 12]
```diff
@@ -18,14 +18,16 @@
...
+    (selectPage)="chooseHomepage()"
     (addPage)="addPage($event)"
   ></api-documentation-home-page-header>
   <mat-card-content>
-    @if (page != null) {
+    @if (homepage != null) {
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.html]:
```suggestion
    @if (homepage) {
```

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts [hunk 13]
```diff
@@ -144,8 +144,8 @@ export class ApiDocumentationV4DefaultPageComponent implements OnInit, OnDestroy
     this.matDialog
       .open<GioConfirmDialogComponent, GioConfirmDialogData, boolean>(GioConfirmDialogComponent, {
         data: {
-          title: 'Delete your page',
-          content: 'Are you sure you want to delete this page? This action is irreversible.',
+          title: 'Confirm Homepage Deletion',
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts]:
```suggestion
          title: 'Delete Homepage',
```

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts [hunk 14]
```diff
@@ -157,12 +157,19 @@ export class ApiDocumentationV4DefaultPageComponent implements OnInit, OnDestroy
       )
       .subscribe({
         next: (_) => {
-          this.snackBarService.success('Page deleted successfully');
+          this.snackBarService.success('Homepage deleted successfully');
           this.ngOnInit();
         },
         error: (error) => {
-          this.snackBarService.error(error?.error?.message ?? 'Error while deleting page');
+          this.snackBarService.error(error?.error?.message ?? 'Error while removing homepage');
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts]:
```suggestion
          this.snackBarService.error(error?.error?.message ?? 'Error when deleting homepage');
```

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts [hunk 15]
```diff
@@ -157,12 +157,19 @@ export class ApiDocumentationV4DefaultPageComponent implements OnInit, OnDestroy
...
         },
       });
   }
+
+  chooseHomepage() {
+    this.router.navigate(['.', 'homepage', 'choose'], {
+      relativeTo: this.activatedRoute,
+      queryParams: { parentId: this.parentId },
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts]: Why do you need the parentId for choosing the homepage?

File: gravitee-apim-console-webui/src/services-ngx/api-documentation-v2.service.ts [hunk 16]
```diff
@@ -38,8 +38,12 @@ export class ApiDocumentationV2Service {
   createDocumentationPage(apiId: string, createDocumentation: CreateDocumentation): Observable<Page> {
     return this.http.post<Page>(`${this.constants.env.v2BaseURL}/apis/${apiId}/pages`, createDocumentation);
   }
-  getApiPages(apiId: string, parentId: string): Observable<ApiDocumentationPageResult> {
-    return this.http.get<ApiDocumentationPageResult>(`${this.constants.env.v2BaseURL}/apis/${apiId}/pages?parentId=${parentId}`).pipe(
+  getApiPages(apiId: string, parentId?: string): Observable<ApiDocumentationPageResult> {
+    const url = parentId
+      ? `${this.constants.env.v2BaseURL}/apis/${apiId}/pages?parentId=${parentId}`
+      : `${this.constants.env.v2BaseURL}/apis/${apiId}/pages`;
+
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/services-ngx/api-documentation-v2.service.ts:45]: If you want, you can do it like this too: ``${this.constants.env.v2BaseURL}/apis/${apiId}/pages${parentId ? '?parentId=' + parentId : ''}``

File: gravitee-apim-console-webui/src/management/api/documentation-v4/api-documentation-v4.module.ts (hunk 1 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/api-documentation-v4.module.ts]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.html (hunk 2 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.html]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.ts (hunk 6 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.ts]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.scss (hunk 7 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.scss]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.ts
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.ts:1]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/documentation-home-page-header/api-documentation-v4-home-page-header.component.html (hunk 10 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/documentation-home-page-header/api-documentation-v4-home-page-header.component.html]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.html (hunk 12 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.html]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts (hunk 14 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts]: Done

File: gravitee-apim-console-webui/src/services-ngx/api-documentation-v2.service.ts (hunk 16 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/services-ngx/api-documentation-v2.service.ts:45]: No, for me that way increases visibility of the difference

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts (hunk 15 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts (hunk 13 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.spec.ts (hunk 9 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.spec.ts]: DOne

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.html (hunk 11 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.html]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.spec.ts (hunk 4 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.spec.ts]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.spec.ts (hunk 8 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-page-list/api-documentation-choose-page-list.component.spec.ts]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.ts (hunk 5 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.ts]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.scss (hunk 3 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.scss]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts [hunk 17]
```diff
@@ -144,8 +144,8 @@ export class ApiDocumentationV4DefaultPageComponent implements OnInit, OnDestroy
     this.matDialog
       .open<GioConfirmDialogComponent, GioConfirmDialogData, boolean>(GioConfirmDialogComponent, {
         data: {
-          title: 'Delete your page',
-          content: 'Are you sure you want to delete this page? This action is irreversible.',
+          title: 'Delete Homepage',
+          content: 'Are you sure you want to delete this page? This action is irreversible. Do you which to continue?',
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts]:
```suggestion
          content: 'Are you sure you want to delete this page? This action is irreversible. Do you want to continue?',
```

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.scss [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    justify-content: center;
+  }
+
+  .header__subtitle {
+    color: mat.m2-get-color-from-palette(gio.$mat-space-palette, 'lighter40');
+    @include mat.m2-typography-level($typography, 'body-2');
+  }
+}
+
+.custom-dialog-content {
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.scss]: Rename without `'dialog'` in the name

File: gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts (hunk 17 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/documentation-default-page/api-documentation-v4-default-page.component.ts]: Done

File: gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.scss (hunk 18 above)
Comment (illichso) [reply gravitee-apim-console-webui/src/management/api/documentation-v4/components/api-documentation-choose-existing-page/api-documentation-choose-existing-page.component.scss]: Done

**Review by jourdiw — APPROVED** [review]
Good to merge once Sonar is happier
